Add real-time barcode scanning API and iOS backend implementation (#4458)
beeware/toga: docs/en/reference/api/hardware/camera.md

    ...
    async def time_for_a_selfie(self, widget, **kwargs):
        photo = await self.camera.take_photo()

    async def scan_qr_code(self, widget, **kwargs):
        content = await self.camera.start_scanning()
        self.label.text = f"Scanned: {content}"
```

Most platforms will require some form of device permission to access the camera. The permission APIs are paired with the specific actions performed on those APIs - that is, to take a photo or scan a barcode, you require [`Camera.has_permission`][toga.hardware.camera.Camera.has_permission], which you can request using [`Camera.request_permission()`][toga.hardware.camera.Camera.request_permission].

Toga will confirm whether the app has been granted permission to use the camera before invoking any camera API. If permission has not yet been granted, the platform *may* request access at the time of first camera access; however, this is not guaranteed to be the behavior on all platforms.

## Scanning for Barcodes

> The camera can be used to scan QR codes and other barcode types in real-time. Scanning is supported on iOS, macOS, and in the Dummy (test) backend.

Reviewer: We don't call out individual platform quirks (or availability) in the docs - they should describe the feature presuming it exists:
```suggestion
The camera can be used to scan QR codes and other barcode types in real-time.
```

    self.log(f"Detected: {content}")

def stop_scan(self, widget, **kwargs):
    self.camera.stop_scanning()
```

You can specify which barcode formats to scan for using the `code_types` parameter. It accepts a single [`BarcodeFormat`][toga.constants.BarcodeFormat] value or a list:

```python
# Scan for QR codes only
content = await self.camera.start_scanning(code_types=BarcodeFormat.QR)

# Scan for multiple formats
content = await self.camera.start_scanning(
    code_types=[BarcodeFormat.QR, BarcodeFormat.CODE128],
)
```

## Notes

- Apps that use a camera must be configured to provide permission to the camera device. The permissions required are platform specific:
    - iOS: `NSCameraUsageDescription` must be defined in the app's `Info.plist` file.
    - macOS: The `com.apple.security.device.camera` entitlement must be enabled, and `NSCameraUsageDescription` must be defined in the app's `Info.plist` file.
    - Android: The `android.permission.CAMERA` permission must be declared.
- The iOS simulator implements the iOS Camera APIs, but is not able to take photographs or scan barcodes. To test your app's Camera usage, you must use a physical iOS device.

> - Barcode scanning is currently available on iOS, macOS, and in the Dummy (test) backend. Other backends will raise `NotImplementedError`.

Reviewer: We don't reference the Dummy backend - it's purely a testing artefact.
```suggestion
- Availability of barcode scanning is currently limited to iOS and macOS.
```
